File: Java/SortArray.java
/*
 * This Java program sorts an unsorted array of integers
 * using multiple different sorting algorithms 
 *
 * Sorting algorithms implemented
 * - Selection Sort
 * - Bubble Sort
 * - Insertion Sort
 * - Quick Sort
 * - TO DO: Merge Sort
 * 
 * Created By: Josh Johnson
 */

import java.util.ArrayList;
import java.util.Scanner;
import java.io.FileNotFoundException;
import java.io.File;
import java.lang.Math;

 public class SortArray {
    public static void main (String [] args) {
        SortArray sa = new SortArray();
        
        /* applying sorting algorithms */
        /* selection sort - size and values are given by program */
        int[] array_one = new int[20];
        array_one = sa.fill_array(array_one);
        sa.print_array(array_one, 1, false, "Selection");
        sa.array_selection_sort(array_one);
        sa.print_array(array_one, 1, true, "Selection");

        /* bubble sort - size and values are given by program */
        int[] array_two = {2, 5, 3, 9, 21, 56, 1, 2, 7, 3, 99, 65, 21, 8};
        sa.print_array(array_two, 2, false, "Bubble");
        sa.array_bubble_sort(array_two);
        sa.print_array(array_two, 2, true, "Bubble");

        /* insertion sort - array size and values are discovered by reading a file */
        ArrayList<Integer> array_three_list = sa.create_array_from_file("../Test_Files/Random_Integers_No_Duplicates.txt");
        int[] array_three = sa.create_array_from_list(array_three_list);
        sa.print_array(array_three, 3, false, "Insertion");
        sa.array_insertion_sort(array_three);
        sa.print_array(array_three, 3, true, "Insertion");

        /* quick sort - array size and values are discovered by reading a file */
        ArrayList<Integer> array_four_list = sa.create_array_from_file("../Test_Files/Random_Integers_With_Duplicates.txt");
        int[] array_four = sa.create_array_from_list(array_four_list);
        sa.print_array(array_four, 4, false, "Quick");
        sa.array_quick_sort(array_four, 0, array_four.length - 1);
        sa.print_array(array_four, 4, true, "Quick");

        /* TO DO: merge sort - array size and values are discovered by reading a file */
    }

    /* reads from a file of integers and adds elements to an arraylist */
    private ArrayList<Integer> create_array_from_file(String file_name) {
        ArrayList<Integer> list = new ArrayList<Integer>();

        try{
            File file_in = new File(file_name);
            Scanner reader = new Scanner(file_in);

            while (reader.hasNextInt()) {
                int value = reader.nextInt();
                list.add(value);
            }

            reader.close();
        } catch (FileNotFoundException e) {
            System.err.println("Could not find file");
            e.printStackTrace();
        }

        return list;
    }

    /* creates an array from an arraylist */
    private int[] create_array_from_list(ArrayList<Integer> list) {
        int arr_size = list.size();
        int[] arr = new int[arr_size];

        for (int i = 0; i < list.size(); i++) {
            arr[i] = list.get(i);
        }

        return arr;
    }

    /* fills in an initialized array with random values */
    private int[] fill_array(int[] arr) {
        int value;

        for (int i = 0; i < arr.length; i++) {
            value = (int) (Math.random()*100) + 1; // random integer 1-100
            arr[i] = value;
        }

        return arr;
    }

    /* performs a selection sort algorithm */
    private int[] array_selection_sort(int[] arr) { 
        /* one by one move boundary of unsorted subarray */
        for (int i = 0; i < (arr.length - 1); i++) {
            /* find the minimum element in unsorted portion array */
            int min_idx = i;
            for (int j = i + 1; j < arr.length; j++) {
                if (arr[j] < arr[min_idx])
                    min_idx = j;
            }
 
            /* swap the found minimum element with the first element */
            if (min_idx != i) {
                int temp = arr[min_idx];
                arr[min_idx] = arr[i];
                arr[i] = temp;
            }
        }

        return arr;
    }

    /* performs a bubble sort algorithm */
    private int[] array_bubble_sort(int[] arr) {
        for (int i = 0; i < arr.length - 1; i++) {
            for (int j = 0; j < arr.length - i - 1; j++) {
                /* swap arr[j+1] and arr[j] */
                if (arr[j] > arr[j + 1]) {
                    int temp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] = temp;
                }
            }
        }

        return arr;
    }

    /* performs an insertion sort algorithm */
    private int[] array_insertion_sort(int[] arr) {
        for (int i = 1; i < arr.length; i++) {
            int key = arr[i];
            int j = i - 1;
 
            /* 
             * move elements of arr[0..i-1], that are greater than key, 
             * to one position ahead of their current position 
             */
            while (j >= 0 && arr[j] > key) {
                arr[j + 1] = arr[j];
                j--;
            }

            arr[j + 1] = key;
        }

        return arr;
    }

    /* performs a quick sort algorithm */
    private int[] array_quick_sort(int[] arr, int low, int high) {
        if (low < high) {
            int index = partition(arr, low, high);
    
            /* separately sort elements before and after partition */
            array_quick_sort(arr, low, index - 1);
            array_quick_sort(arr, index + 1, high);
        }

        return arr;
    }

    /* 
     * takes the last element as pivot, places the pivot element at 
     * correct position in sorted array, and places all smaller elements
     * to the left of the pivot and all greater elements to right of pivot 
     */
    private int partition(int[] arr, int low, int high) {
        int temp;
        int pivot = arr[high];
        int i = (low - 1);
    
        for(int j = low; j <= high - 1; j++) {
            if (arr[j] < pivot) {
                i++;

                /* swap two elements */
                temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        /* swap two elements */
        temp = arr[i+1];
        arr[i+1] = arr[high];
        arr[high] = temp;

        return (i + 1);
    }
    
    /* prints the contents of the given integer array to stdout */
    private void print_array(int[] arr, int arr_id, boolean ordered, String sorting_type) {
        if (ordered){
            System.out.print("Array with ID " + arr_id 
            + " after " + sorting_type + " Sort algorithm applied: ");
        }
        else {
            System.out.print("Array with ID " + arr_id 
            + " before " + sorting_type + " Sort algorithm applied: ");
        }

        for (int i = 0; i < arr.length; i++) {
            if (i == (arr.length - 1)) {
                System.out.println(arr[i]);
                break;
            }

            System.out.print(arr[i] + " ");
        }
    }
 }

File: Python/SortArray.py
# This Python program sorts unsorted arrays of integers given by
# the program itself, as well as given through text files
# using multiple different sorting algorithms 
#
# Sorting algorithms implemented
# - Bubble
# - Selection
# - Merge
# - Insertion
# - Quick
#
# Created By: Josh Johnson

from os import path # used to get relative path for text files
import random # used for random value generation

# bubble sort algorithm
def bubble_sort(array):
    # if the array is already sorted
    swapped = False

    # traverse through all array elements
    for i in range(len(array) - 1):
        # previous i elements are already in place
        for j in range(0, len(array) - i - 1):
            # swap if the current element is greater than the next element
            if array[j] > array[j + 1]:
                swapped = True
                array[j], array[j + 1] = array[j + 1], array[j]
         
        if not swapped:
            return

# selection sort algorithm
def selection_sort(array):
    array_size = len(array)

    # traverse through all array elements
    for i in range(array_size):
        min_index = i

        # select the minimum element in every iteration
        for j in range(i + 1, array_size):
            if array[j] < array[min_index]:
                min_index = j

         # swapping the elements to sort the array if necessary
        if array[i] != array[min_index]:
            (array[i], array[min_index]) = (array[min_index], array[i])

# merge sort algorithm
def merge_sort(array, left_index, right_index):
    if left_index < right_index:
        middle_index = left_index + (right_index - left_index) // 2 # // results in integer division rather than / being double division
 
        # sort first and second halves of sub arrays
        merge_sort(array, left_index, middle_index)
        merge_sort(array, middle_index + 1, right_index)
        merge(array, left_index, middle_index, right_index)

# merges two subarrays of the array
def merge(array, left_index, middle_index, right_index):
    # get sizes of sub arrays
    subarray1_size = middle_index - left_index + 1
    subarray2_size = right_index - middle_index
 
    # create temp arrays
    temp_array1 = [0] * subarray1_size
    temp_array2 = [0] * subarray2_size
 
    # copy data to temp_array1
    for i in range(subarray1_size):
        temp_array1[i] = array[left_index + i]

    # copy data to temp_array2
    for i in range(subarray2_size):
        temp_array2[i] = array[middle_index + 1 + i]

    i, j, k = 0, 0, left_index # multiple variables on the same line
 
    # merge two sorted sub-arrays into one sorted array
    while i < subarray1_size and j < subarray2_size:
        if temp_array1[i] <= temp_array2[j]:
            array[k] = temp_array1[i]
            i += 1
        else:
            array[k] = temp_array2[j]
            j += 1
            
        k += 1
 
    # copy the remaining elements of temp_array1, if there are any
    while i < subarray1_size:
        array[k] = temp_array1[i]
        i += 1
        k += 1
 
    # copy the remaining elements of temp_array2, if there are any
    while j < subarray2_size:
        array[k] = temp_array2[j]
        j += 1
        k += 1

# insertion sort algorithm
def insertion_sort(array):
    for i in range(1, len(array)): 
        key = array[i]
 
        # move elements of arr[0..i-1], that are greater than key, 
        # to one position ahead of their current position
        j = i - 1
        while j >= 0 and key < array[j]:
                array[j + 1] = array[j]
                j -= 1

        array[j + 1] = key

# quick sort algorithm
#   - this implementation utilizes pivot as the last element in the array
def quick_sort(array, low, high):
    if low < high:
        pivot_index = partition(array, low, high)
 
        # recursive call on the left and right of pivot
        quick_sort(array, low, pivot_index - 1) 
        quick_sort(array, pivot_index + 1, high) 
 
# finds the partition position
def partition(array, low, high):
    pivot = array[high] # set the right-most element as pivot
    i = low - 1 # pointer for greater element
 
    # traverse through all elements comparing each element with pivot
    for j in range(low, high):
        # if element smaller than pivot is found swap it with the greater element pointed by i
        if array[j] <= pivot:
            i += 1
 
            # swapping element at i with element at j if necessary
            if array[i] != array[j]:
                (array[i], array[j]) = (array[j], array[i])
 
    # swap the pivot element with the greater element specified by i if necessary
    if array[i + 1] != array[high]:
        (array[i + 1], array[high]) = (array[high], array[i + 1])

    return i + 1 # position from where partition is done
 
# generates an array with random integers as elements with given array size
def create_random_array(array_size):
    array = []

    for _ in range(array_size):
        element = random.randint(-100, 100) # random value between range; inclusive
        array.append(element)

    return array

# reads a test file and returns an array of elements gathered by the file separated by a space
def read_test_file_into_array(file_name):
    # reading from file
    base_path = path.dirname(__file__) # get current directory location 
    file_path = path.abspath(path.join(base_path, "..", "Test_Files", file_name)) # relative path to the test file
    file1 = open(file_path, "r") # open access to test file
    data = file1.read() # storing file content
    file1.close() # closing the file access to test file

    # creating array from content gathered from test file
    array = [int(i) for i in data.split(' ')] # create array from elements separated by a space in data variable
    return array
     
# prints the contents of the given integer array to stdout
def print_array(array, array_id, ordered, sorting_algorithm):
    if ordered:
        print("Array with ID %d after the %s Sort algorithm is applied:" % (array_id, sorting_algorithm), end=" ") # end parameter prevents new line
    else:
        print("Array with ID %d before the %s Sort algorithm is applied:" % (array_id, sorting_algorithm), end=" ") # end parameter prevents new line
    
    for i in range(len(array)):
        if i == len(array) - 1:
            print("%d" % array[i])
        else:
            print("%d" % array[i], end=", ")

# main function - controls flow of the program
def main():
    array_id = 0 # array ID tracker
    arrays = [0] * 5 # multiplied by 5 for number of algorithms implemented initialized to 0

    # apply sorting algorithms
    # bubble sort - size and values are given by program
    array_one = [3, 66, 45, 23, 9, 0, 11, 2, 7, 49, 21]
    arrays[array_id] = array_one

    print_array(arrays[array_id], array_id, False, "Bubble")
    bubble_sort(arrays[array_id])
    print_array(arrays[array_id], array_id, True, "Bubble")

    # selection sort - size and values are given by program
    array_id += 1
    arrays[array_id] = [55, 4, 32, 62, -2, 4, 6, 9, 99, 12, -12, -5, 4, 55]

    print_array(arrays[array_id], array_id, False, "Selection")
    selection_sort(arrays[array_id])
    print_array(arrays[array_id], array_id, True, "Selection")

    # merge sort - size and values are given by program
    array_id += 1
    arrays[array_id] = create_random_array(10)

    print_array(arrays[array_id], array_id, False, "Merge")
    merge_sort(arrays[array_id], 0, len(arrays[array_id]) - 1)
    print_array(arrays[array_id], array_id, True, "Merge")

    # insertion sort - array size and values are discovered by reading a file
    array_id += 1
    arrays[array_id] = read_test_file_into_array("random_integers_no_duplicates.txt")

    print_array(arrays[array_id], array_id, False, "Insertion")
    insertion_sort(arrays[array_id])
    print_array(arrays[array_id], array_id, True, "Insertion")

    # quick sort - array size and values are discovered by reading a file
    array_id += 1
    arrays[array_id] = read_test_file_into_array("random_integers_with_duplicates.txt")

    print_array(arrays[array_id], array_id, False, "Quick")
    quick_sort(arrays[array_id], 0, len(arrays[array_id]) - 1)
    print_array(arrays[array_id], array_id, True, "Quick")
    
# begins program by calling the main function
main()